NMS-11065: WmiMonitor (#4731)
OpenNMS/opennms: docs/modules/reference/pages/service-assurance/monitors/WmiMonitor.adoc

= WmiMonitor

The WMI monitor tests the availability of WMI services on remote interfaces.

NOTE: xref:reference:service-assurance/monitors/WsManMonitor.adoc[WSMAN] does everything that WMI does and is a more modern protocol for Windows monitoring.

== Monitor facts

[cols="1,7"]
|===
| Class Name
| `org.opennms.netmgt.poller.monitors.WmiMonitor`
|===

== Configuration and use

> Note that username, password, and domain credentials defined here override any defined in the wmi-config file.

Reviewer: @Bonrob2 Is it required to have the credentials within the wmi-config file? If not MetaData's are also an option to avoid credentials in poller configs. But maybe this is a general hint we could add in the common section.

Reviewer 2: It's not required, though if using directory credentials, using the config file can apply a single credential to a range of devices rather than storing individual credentials in metadata and/or service configs.

= WmiMonitor

The WMI monitor tests the availability of WMI services on remote interfaces.

> Note that WSMAN does everything that WMI does and is a more modern protocol for Windows monitoring.

Reviewer 2: ```suggestion
NOTE: xref:reference:service-assuranve/monitors/WsManMonitor.adoc[WSMAN] does everything that WMI does and is a more modern protocol for Windows monitoring.
```

//verify
This monitor implements <<reference:service-assurance/introduction.adoc#ref-service-assurance-monitors-placeholder-substitution-parameters, placeholder substitution in parameter values>>.

== Monitor facts

[cols="1,7"]
|===
| Class Name
| `org.opennms.netmgt.poller.monitors.WmiMonitor`
|===

== Configuration and use

.Monitor-specific parameters for the WmiMonitor
[options="header"]
[cols="1,3,2"]
|===
| Parameter
| Description
| Default

3+| *Required*

| username
| The username to use when authenticating against the WMI host.
| none

| password
| The password to use when authenticating against the WMI host.
| none

| domain
| Windows domain where the user is located.
You do not have to use the domain parameter if you use local user accounts.

> | none

Reviewer 2: These are only required if the user has not defined credentials in `$\{OPENNMS_HOME}/etc/wmi-config.xml`
Credentials given here will override any defined in the wmi-config file.  It's probably a better practice to put credentials there than in the pollerd/collectd config.

| Tells the OpenNMS WMI system to deal with multiple instance results.
Available match types are:

* *all*: all instances must comply to the compare operation.
* *none*: no instances should comply with the compare operation.
* *some*: only some (one or more) must comply with the compare operation.
* *one*: only one instance can comply with the compare operation.
| all

| wmiClass
| Class of instances within WMI.
This class contains objects representing properties and methods available.
| Win32_ComputerSystem

| wmiObject
| Member of a WMI property set.
For example, the `Win32_ComputerSystem` class contains properties that we can look at and poll.
The default "WMI" service uses the "Status" property to determine if the system is running and if WMI is available.
| Status

| compareOp
| The compare operation tells the OpenNMS WMI poller plugins how to verify the nature (up, critical) of a specific property within WMI.
In the event of unequal comparisons, the system will always have the WMI value (as retrieved from the target) on the left.
Available compare operations are:

* *EQ* (equals)
* *NEQ* (not equals)
* *GT* (greater than)
* *LT* (less than)
* *NOOP* (no operation)

> //not sure what "manager" is referring to here.
> 
> The NOOP operation is a special case in which the Manager always returns a result code of "OK."

Reviewer 2: ```suggestion
The NOOP operation is a special case in which always evaluates to a result code of "OK."
```

This is typically used in scenarios where the existence of a class or property is more interesting than the contents of the property.

> OpenNMS uses this in the collector to check and ensure that WMI is available before attempting to collect data.

Reviewer 2: This is good-to-know info, though not specifically relevant to the poller.

| EQ

| compareValue
| The value to use when performing a comparison against the wmiObject.
| OK

| service-name
| Name of the service to monitor.

>  ##If there is a ~ at the beginning of the parameter, the rest of the string will be used as a regular expression pattern match, otherwise it will be considered an exact match.##

Reviewer 2: Does this need the `##` around it?

Author: I put that there instead of a comment, as I wanted to make sure it was correct.

| WMI

| retry
| Number of retries before the service is marked as down.
| 0

| timeout
| Time in milliseconds to wait for a response.
|

| namespace
| Every WMI class exists in a namespace.
The default namespace for most common classes is root/cimv2.
It is possible to refer to WMI objects from other namespaces such as root/MicrosoftActiveDirectory.
| root/cimv2

|===

kbd:[{}] _indicates the parameter supports <<reference:service-assurance/introduction.adoc#ref-service-assurance-monitors-placeholder-substitution-parameters, placeholder substitution>>._

This monitor implements the <<reference:service-assurance/introduction.adoc#ref-service-assurance-monitors-common-parameters, Common Configuration Parameters>>.

== Example

> The following example shows how to enable monitoring that already discovered WMI services are still running.

Reviewer 2: This is a valid example, though it seems like a circular check to use WMI to see if the WMI service is running.  If the service is down, the poller will fail due to not being able to read data, not that it read a value of down. :D

Author: LOL. Makes sense. I just copied it from somewhere else. Could you suggest a better example?
